Building with Xtensa toolchain fails

This worked until very recently, now I get:
```
$ west build -b intel_s1000_crb samples/hello_world
...
[8/93] Building C object zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj
FAILED: zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj 
ccache /home/jh/xtensa/XtDevTools/install/tools/RF-2016.4-linux/XtensaTools/bin/xt-xcc -DBUILD_VERSION=zephyr-v1.14.0-1759-gc790d5b4982b -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I../kernel/include -I../arch/xtensa/include -I../include -I../include/drivers -Izephyr/include/generated -I../soc/xtensa/intel_s1000 -I../soc/xtensa/intel_s1000/include -I../lib/libc/minimal/include -Os --imacros=/home/jh/src/zephyr/zephyr/build/zephyr/include/generated/autoconf.h --imacros=/home/jh/src/zephyr/zephyr/include/toolchain/zephyr_stdint.h -ffreestanding -fno-common -g -fms-extensions -D__SIZEOF_LONG__=4 -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Werror=implicit-int -fno-pic -fno-strict-overflow -ffunction-sections -fdata-sections -mlongcalls -std=c99 -MD -MT zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/xtensa/core/offsets/offsets.c.obj   -c /home/jh/src/zephyr/zephyr/arch/xtensa/core/offsets/offsets.c
cc1: error: =/home/jh/src/zephyr/zephyr/build/zephyr/include/generated/autoconf.h: No such file or directory
cc1: error: =/home/jh/src/zephyr/zephyr/include/toolchain/zephyr_stdint.h: No such file or directory
In file included from ../include/toolchain/xcc.h:10,
                 from ../include/toolchain.h:37,
                 from ../kernel/include/gen_offset.h:74,
                 from /home/jh/src/zephyr/zephyr/arch/xtensa/core/offsets/offsets.c:26:
../include/toolchain/gcc.h:371:2: error: #error processor architecture not supported
In file included from ../include/kernel_includes.h:34,
                 from ../include/kernel.h:17,
                 from ../kernel/include/kernel_structs.h:10,
                 from /home/jh/src/zephyr/zephyr/arch/xtensa/core/offsets/offsets.c:30:
../include/arch/cpu.h:29:2: error: #error "Unknown Architecture"
In file included from ../kernel/include/kernel_structs.h:10,
                 from /home/jh/src/zephyr/zephyr/arch/xtensa/core/offsets/offsets.c:30:
../include/kernel.h:48:2: error: #error "invalid configuration"
../include/kernel.h:64:2: error: #error "invalid configuration"
```

If I rebase back to a few days old tree everything works. I suspect #16645 might be the culprit.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -220,9 +220,6 @@ toolchain_cc_nostdinc()
 # @Intent: Set compiler specific macro inclusion of AUTOCONF_H
 toolchain_cc_imacros(${AUTOCONF_H})
 
-# @Intent: Enforce standard integer type correspondance to match Zephyr usage.
-toolchain_cc_imacros(${ZEPHYR_BASE}/include/toolchain/zephyr_stdint.h)
-
 # @Intent: Set compiler specific flag for bare metal freestanding option
 toolchain_cc_freestanding()
 
@@ -239,6 +236,10 @@ toolchain_cc_asm_base_flags(ASM_BASE_FLAG)
 zephyr_compile_options(
   $<$<COMPILE_LANGUAGE:ASM>:${ASM_BASE_FLAG}>
 )
+
+# @Intent: Enforce standard integer type correspondance to match Zephyr usage.
+# (must be after compiler specific flags)
+toolchain_cc_imacros(${ZEPHYR_BASE}/include/toolchain/zephyr_stdint.h)
 
 # Common toolchain-agnostic assembly flags
 zephyr_compile_options(

--- a/cmake/compiler/gcc/target_imacros.cmake
+++ b/cmake/compiler/gcc/target_imacros.cmake
@@ -4,6 +4,12 @@
 
 macro(toolchain_cc_imacros header_file)
 
-  zephyr_compile_options(--imacros=${header_file})
+  # We cannot use the "-imacros foo" form here as CMake insists on
+  # deduplicating arguments, meaning that subsequent usages after the
+  # first one will see the "-imacros " part removed.
+  # gcc and clang support the "--imacros=foo" form but not xcc.
+  # Let's use the "combined" form (without space) which is supported
+  # by everyone so far.
+  zephyr_compile_options(-imacros${header_file})
 
 endmacro()

--- a/cmake/toolchain/xcc/generic.cmake
+++ b/cmake/toolchain/xcc/generic.cmake
@@ -29,10 +29,8 @@ set(NOSYSDEF_CFLAG "")
 
 list(APPEND TOOLCHAIN_C_FLAGS -fms-extensions)
 
-# xcc doesn't have this, so we need to define it here.
-# This is the same as in the xcc toolchain spec files.
 list(APPEND TOOLCHAIN_C_FLAGS
-  -D__SIZEOF_LONG__=4
+  -imacros${ZEPHYR_BASE}/include/toolchain/xcc_missing_defs.h
   )
 
 set(TOOLCHAIN_HAS_NEWLIB OFF CACHE BOOL "True if toolchain supports newlib")

--- a/include/toolchain/xcc_missing_defs.h
+++ b/include/toolchain/xcc_missing_defs.h
@@ -1,0 +1,170 @@
+/*
+ * Copyright (c) 2019 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Basic macro definitions that gcc and clang provide on their own
+ * but that xcc lacks. Only those that Zephyr requires are provided here.
+ */
+
+#ifndef ZEPHYR_INCLUDE_TOOLCHAIN_XCC_MISSING_DEFS_H_
+#define ZEPHYR_INCLUDE_TOOLCHAIN_XCC_MISSING_DEFS_H_
+
+#if __CHAR_BIT__ == 8
+#define __SCHAR_WIDTH__ 8
+#else
+#error "unexpected __CHAR_BIT__ value"
+#endif
+
+#if __SHRT_MAX__ == 32767
+#define __SHRT_WIDTH__ 16
+#define __SIZEOF_SHORT__ 2
+#else
+#error "unexpected __SHRT_WIDTH__ value"
+#endif
+
+#if __INT_MAX__ == 2147483647
+#define __INT_WIDTH__ 32
+#define __SIZEOF_INT__ 4
+#else
+#error "unexpected __INT_MAX__ value"
+#endif
+
+#if __LONG_MAX__ == 2147483647L
+#define __LONG_WIDTH__ 32
+#define __SIZEOF_LONG__ 4
+#else
+#error "unexpected __LONG_MAX__ value"
+#endif
+
+#if __LONG_LONG_MAX__ == 9223372036854775807LL
+#define __LONG_LONG_WIDTH__ 64
+#define __SIZEOF_LONG_LONG__ 8
+#else
+#error "unexpected __LONG_LONG_MAX__ value"
+#endif
+
+#if __INTMAX_MAX__ == 9223372036854775807LL
+#define __INTMAX_WIDTH__ 64
+#define __SIZEOF_INTMAX__ 8
+#define __UINTMAX_MAX__ 0xffffffffffffffffULL
+#define __UINTMAX_WIDTH__ 64
+#define __SIZEOF_UINTMAX__ 8
+#else
+#error "unexpected __INTMAX_MAX__ value"
+#endif
+
+/*
+ * No xcc provided definitions related to pointers, so let's just enforce
+ * the Zephyr expected type.
+ */
+
+#define __INTPTR_MAX__ 0x7fffffffL
+#define __INTPTR_TYPE__ long int
+#define __INTPTR_WIDTH__ 32
+#define __SIZEOF_POINTER__ 4
+
+#define __PTRDIFF_MAX__ 0x7fffffffL
+#define __PTRDIFF_WIDTH__ 32
+#define __SIZEOF_PTRDIFF_T__ 4
+
+#define __UINTPTR_MAX__ 0xffffffffLU
+#define __UINTPTR_TYPE__ long unsigned int
+
+/*
+ * xcc already defines __SIZE_TYPE__ as "unsigned int" but there is no way
+ * to safeguard that here with preprocessor equality.
+ */
+
+#define __SIZE_MAX__ 0xffffffffU
+#define __SIZE_WIDTH__ 32
+#define __SIZEOF_SIZE_T__ 4
+
+/*
+ * The following defines are inferred from the xcc provided defines
+ * already tested above.
+ */
+
+#define __INT8_MAX__ 0x7f
+#define __INT8_TYPE__ signed char
+
+#define __INT16_MAX__ 0x7fff
+#define __INT16_TYPE__ short int
+
+#define __INT32_MAX__ 0x7fffffff
+#define __INT32_TYPE__ int
+
+#define __INT64_MAX__ 0x7fffffffffffffffLL
+#define __INT64_TYPE__ long long int
+
+#define __INT_FAST8_MAX__ 0x7f
+#define __INT_FAST8_TYPE__ signed char
+#define __INT_FAST8_WIDTH__ 8
+
+#define __INT_FAST16_MAX__ 0x7fffffff
+#define __INT_FAST16_TYPE__ int
+#define __INT_FAST16_WIDTH__ 32
+
+#define __INT_FAST32_MAX__ 0x7fffffff
+#define __INT_FAST32_TYPE__ int
+#define __INT_FAST32_WIDTH__ 32
+
+#define __INT_FAST64_MAX__ 0x7fffffffffffffffLL
+#define __INT_FAST64_TYPE__ long long int
+#define __INT_FAST64_WIDTH__ 64
+
+#define __INT_LEAST8_MAX__ 0x7f
+#define __INT_LEAST8_TYPE__ signed char
+#define __INT_LEAST8_WIDTH__ 8
+
+#define __INT_LEAST16_MAX__ 0x7fff
+#define __INT_LEAST16_TYPE__ short int
+#define __INT_LEAST16_WIDTH__ 16
+
+#define __INT_LEAST32_MAX__ 0x7fffffff
+#define __INT_LEAST32_TYPE__ int
+#define __INT_LEAST32_WIDTH__ 32
+
+#define __INT_LEAST64_MAX__ 0x7fffffffffffffffLL
+#define __INT_LEAST64_TYPE__ long long int
+#define __INT_LEAST64_WIDTH__ 64
+
+#define __UINT8_MAX__ 0xff
+#define __UINT8_TYPE__ unsigned char
+
+#define __UINT16_MAX__ 0xffff
+#define __UINT16_TYPE__ short unsigned int
+
+#define __UINT32_MAX__ 0xffffffffU
+#define __UINT32_TYPE__ unsigned int
+
+#define __UINT64_MAX__ 0xffffffffffffffffULL
+#define __UINT64_TYPE__ long long unsigned int
+
+#define __UINT_FAST8_MAX__ 0xff
+#define __UINT_FAST8_TYPE__ unsigned char
+
+#define __UINT_FAST16_MAX__ 0xffffffffU
+#define __UINT_FAST16_TYPE__ unsigned int
+
+#define __UINT_FAST32_MAX__ 0xffffffffU
+#define __UINT_FAST32_TYPE__ unsigned int
+
+#define __UINT_FAST64_MAX__ 0xffffffffffffffffULL
+#define __UINT_FAST64_TYPE__ long long unsigned int
+
+#define __UINT_LEAST8_MAX__ 0xff
+#define __UINT_LEAST8_TYPE__ unsigned char
+
+#define __UINT_LEAST16_MAX__ 0xffff
+#define __UINT_LEAST16_TYPE__ short unsigned int
+
+#define __UINT_LEAST32_MAX__ 0xffffffffU
+#define __UINT_LEAST32_TYPE__ unsigned int
+
+#define __UINT_LEAST64_MAX__ 0xffffffffffffffffULL
+#define __UINT_LEAST64_TYPE__ long long unsigned int
+
+#endif

